mdbook-spec/src/grammar/render_railroad.rs

```diff
@@ -0,0 +1,283 @@
+                // sure.
+                e.render_railroad(stack, link_map, reverse)?
+            }
+            ExpressionKind::Alt(es) => {
+                let choices: Vec<_> = es
+                    .iter()
+                    .map(|e| e.render_railroad(stack, link_map, reverse))
+                    .filter_map(|n| n)
+                    .collect();
+                Box::new(Choice::<Box<dyn Node>>::new(choices))
+            }
+            ExpressionKind::Sequence(es) => {
+                let mut es: Vec<_> = es.iter().collect();
+                // For reversing, see ::Repeat for an explanation.
+                if reverse {
+                    es.reverse();
+                }
+                let make_seq = |es: &[&Expression]| {
+                    let seq: Vec<_> = es
+                        .iter()
+                        .map(|e| e.render_railroad(stack, link_map, reverse))
+                        .filter_map(|n| n)
+                        .collect();
+                    let seq: Sequence<Box<dyn Node>> = Sequence::new(seq);
+                    Box::new(seq)
+                };
+
+                // If `stack` is true, split the sequence on Breaks and stack them vertically.
+                if stack {
+                    // First, trim a Break from the front and back.
+                    let es = if matches!(
+                        es.first(),
+                        Some(e) if e.is_break()
+                    ) {
+                        &es[1..]
+                    } else {
+                        &es[..]
+                    };
+                    let es = if matches!(
+                        es.last(),
+                        Some(e) if e.is_break()
+                    ) {
+                        &es[..es.len() - 1]
+                    } else {
+                        &es[..]
+                    };
+
+                    let mut breaks: Vec<_> =
+                        es.split(|e| e.is_break()).map(|es| make_seq(es)).collect();
+                    // If there aren't any breaks, don't bother stacking.
+                    if breaks.len() == 1 {
+                        breaks.pop().unwrap()
+                    } else {
+                        Box::new(Stack::new(breaks))
+                    }
+                } else {
+                    make_seq(&es)
+                }
+            }
+            ExpressionKind::Optional(e) => {
+                let n = e.render_railroad(stack, link_map, reverse)?;
+                Box::new(Optional::new(n))
+            }
+            ExpressionKind::Repeat(e) => {
+                // Railroad renders everything in the opposite order. However,
+                // our grammar is not written that way, so we need to undo the
+                // reversal.
+                let n = e.render_railroad(stack, link_map, !reverse)?;
+                Box::new(Repeat::new(railroad::Empty, n))
+            }
```

Comment:
The alternative to this reversing would be to render the `n` on the main forward line, for the loop to be `Empty`, and then to render it as `Optional`.

I'm not sure which is better.  The `Optional` way is a bit noisier, but on the other hand, it keeps the grammar in the forward direction, which is nice.

Reply:
Made this (and the change to the one below) in a commit.